crates/wasmtime/src/store.rs

```diff
@@ -102,251 +156,95 @@ enum OutOfGas {
-            let new = slot.get().saturating_add(amt);
-            if new > max {
-                bail!(
-                    "resource limit exceeded: {} count too high at {}",
-                    desc,
-                    new
-                );
-            }
-            slot.set(new);
-            Ok(())
-        }
-
-        let module = module.env_module();
-        let memories = module.memory_plans.len() - module.num_imported_memories;
-        let tables = module.table_plans.len() - module.num_imported_tables;
-        let (max_instances, max_memories, max_tables) = self.limits();
-
-        bump(&self.inner.instance_count, max_instances, 1, "instance")?;
-        bump(&self.inner.memory_count, max_memories, memories, "memory")?;
-        bump(&self.inner.table_count, max_tables, tables, "table")?;
-
-        Ok(())
-    }
-
-    pub(crate) unsafe fn add_instance(
-        &self,
-        handle: InstanceHandle,
-        ondemand: bool,
-    ) -> StoreInstanceHandle {
-        self.inner.instances.borrow_mut().push(StoreInstance {
-            handle: handle.clone(),
-            ondemand,
+                current_poll_cx: UnsafeCell::new(ptr::null_mut()),
+            },
+            out_of_gas_behavior: OutOfGas::Trap,
+            store_data: StoreData::new(),
+            limiter: None,
+            default_callee,
+            data,
         });
-        StoreInstanceHandle {
-            store: self.clone(),
-            handle,
-        }
-    }
 
-    pub(crate) fn existing_instance_handle(&self, handle: InstanceHandle) -> StoreInstanceHandle {
-        debug_assert!(
-            self.inner
-                .instances
-                .borrow()
-                .iter()
-                .any(|i| i.handle.vmctx_ptr() == handle.vmctx_ptr())
-                || self.inner.engine.host_func_anyfunc(&handle).is_some()
-        );
-        StoreInstanceHandle {
-            store: self.clone(),
-            handle,
+        // Once we've actually allocated the store itself we can configure the
+        // trait object pointer of the default callee.
+        let store = StoreContextMut(&mut *inner).opaque().traitobj;
+        unsafe {
+            inner.default_callee.set_store(store);
         }
+        Self { inner }
     }
 
-    pub(crate) unsafe fn existing_vmctx(&self, cx: *mut VMContext) -> StoreInstanceHandle {
-        self.existing_instance_handle(InstanceHandle::from_vmctx(cx))
+    /// Access the underlying data owned by this `Store`.
+    pub fn data(&self) -> &T {
+        self.inner.data()
     }
 
-    #[cfg_attr(not(target_os = "linux"), allow(dead_code))] // not used on all platforms
-    pub(crate) fn set_signal_handler(&self, handler: Option<Box<SignalHandler<'static>>>) {
-        *self.inner.signal_handler.borrow_mut() = handler;
+    /// Access the underlying data owned by this `Store`.
+    pub fn data_mut(&mut self) -> &mut T {
+        self.inner.data_mut()
```

Comment:
can we additionally have an `into_inner(self) -> T`?

Reply:
Ah yeah I'd forgotten to come back and implement this, now implemented though!